proto/table_marshal.go

```diff
@@ -167,6 +168,27 @@ func (u *marshalInfo) size(ptr pointer) int {
 		u.computeMarshalInfo()
 	}
 
+	// Uses the message's Size method if available
+	if u.hassizer {
+		// need to call the sizer of all other fields in order
+		// to compute their marshalinfo
+		for _, f := range u.fields {
+			if f.isPointer && ptr.offset(f.field).getPointer().isNil() {
+				// nil pointer always marshals to nothing
+				continue
+			}
+			f.sizer(ptr.offset(f.field), f.tagsize)
+		}
+
+		s := ptr.asPointerTo(u.typ).Interface().(Sizer)
+		n := s.Size()
```

Comment:
As discussed, now we are now effectively calculating the size twice.

Reply:
Yeh. But the current size recursively calls size again if needed and that initialized struct field's marshalinfos, calculate the struct field size and store it at the cache location. If we dont call size on struct fields we don't cache their sizes. unfortunate the sizerers don't really need this recursive size call.
We could check if the thing is a marshaler and sizer it can just return the size, because we dont need the caching then. But other than that we do need to be able to cache the size for the default marshaler. Hmm. I might have another idea.